File: lab7 - całkowanie-numeryczne-1/romberg.m
function result = romberg_integration(f, a, b, tol)
    if nargin < 4
        tol = 1e-6;
    end
    R = zeros(10, 10);
    R(1,1) = 0.5 * (b - a) * (f(a) + f(b));
    n = 1;
    i = 1;
    while true
        h = (b - a) / (2^n);
        sum_f = sum(arrayfun(f, a + (2*(1:2^(n-1)) - 1) * h));
        R(i+1,1) = 0.5 * R(i,1) + sum_f * h;
        for j = 2:i+1
            R(i+1,j) = R(i+1,j-1) + (R(i+1,j-1) - R(i,j-1)) / (4^(j-1) - 1);
        end
        if abs(R(i+1,i+1) - R(i,i)) < tol
            break;
        end
        i = i + 1;
        n = n + 1;
    end
    result = R(i+1,i+1);
end

% Przykład użycia
f = @(x) x.^2;
a = 0;
b = 1;
result = romberg_integration(f, a, b);
disp(['Przybliżona wartość całki metodą Romberga: ', num2str(result)])

File: lab7 - całkowanie-numeryczne-1/romberg.py
def romberg_integration(f, a, b, tol=1e-6):
    def trapezoid(f, a, b, n):
        h = (b - a) / n
        integral = (f(a) + f(b)) / 2.0
        for i in range(1, n):
            x = a + i * h
            integral += f(x)
        integral *= h
        return integral

    R = [[0.5 * (b - a) * (f(a) + f(b))]]  # R(0, 0)
    n = 1
    i = 0
    while True:
        h = (b - a) / (2 ** n)
        sum_f = sum(f(a + (2 * k - 1) * h) for k in range(1, 2**(n-1) + 1))
        R_i_0 = 0.5 * R[i][0] + sum_f * h
        R.append([R_i_0])
        for j in range(1, i + 2):
            R_ij = R[i + 1][j - 1] + (R[i + 1][j - 1] - R[i][j - 1]) / (4**j - 1)
            R[i + 1].append(R_ij)
        if abs(R[i + 1][i + 1] - R[i][i]) < tol:
            break
        i += 1
        n += 1
    return R[i + 1][i + 1]

# Przykład użycia
f = lambda x: x**2
a = 0
b = 1
result = romberg_integration(f, a, b)
print(f"Przybliżona wartość całki metodą Romberga: {result}")
